contracts/src/receipt.rs

```diff
@@ -0,0 +1,109 @@
+// Copyright 2025 RISC Zero, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use alloy_primitives::Bytes;
+use risc0_aggregation::{
+    decode_set_inclusion_seal, SetInclusionDecodingError, SetInclusionEncodingError,
+    SetInclusionReceipt,
+};
+use risc0_zkvm::{FakeReceipt, InnerReceipt, Receipt as Risc0Receipt, ReceiptClaim};
+use thiserror::Error;
+
+use crate::{
+    encode_seal,
+    groth16::decode_groth16_seal,
+    selector::{SelecorType, Selector, SelectorError},
+};
+
+/// Wrapper around different types of receipts.
+pub enum ReceiptType {
+    Risc0(Risc0Receipt),
+    SetInclusion(SetInclusionReceipt<ReceiptClaim>),
+}
+
+impl ReceiptType {
+    /// Encode the receipt as a seal.
+    pub fn encode_seal(&self) -> Result<Vec<u8>, SetInclusionEncodingError> {
+        match self {
+            ReceiptType::Risc0(receipt) => {
+                encode_seal(receipt).map_err(|_| SetInclusionEncodingError::UnsupportedReceiptType)
+            }
+            ReceiptType::SetInclusion(receipt) => receipt.abi_encode_seal(),
+        }
+    }
+
+    /// Get the receipt if it is a RISC0 receipt.
+    pub fn receipt(&self) -> Option<Risc0Receipt> {
+        match self {
+            ReceiptType::Risc0(receipt) => Some(receipt.clone()),
+            _ => None,
+        }
+    }
+
+    /// Get the receipt if it is a set inclusion receipt.
+    pub fn set_inclusion_receipt(&self) -> Option<&SetInclusionReceipt<ReceiptClaim>> {
+        match self {
+            ReceiptType::SetInclusion(receipt) => Some(receipt),
+            _ => None,
+        }
+    }
+}
+
+/// Errors that can occur when decoding a seal.
+#[derive(Debug, Error)]
+#[non_exhaustive]
+pub enum DecodingError {
+    #[error("Seal too short")]
+    SealTooShort,
+    #[error("Unsupported selector {0:?}")]
+    UnsupportedSelector([u8; 4]),
+    #[error("Selector error: {0}")]
+    SelectorError(#[from] SelectorError),
+    #[error("Decoding error: {0}")]
+    SetInclusionError(#[from] SetInclusionDecodingError),
+    #[error("Decoding error: {0}")]
+    Anyhow(#[from] anyhow::Error),
+}
+
+/// Decode a seal into a receipt.
+pub fn decode_seal(
```

Comment:
I'd recommend we split this into two function APIs. We can have an `fn decode_seal(seal: Bytes, image_id: impl Into<Digest>, journal: impl Into<Vec<u8>>)` and an `fn decode_seal_with_claim(seal: Bytes, claim: ReceiptClaim, journal: impl Into<Vec<u8>>)`. Reason being, most users don't ever use `ReceiptClaim` directly or really know much about it. Given an image ID and journal, we can call `ReceiptClaim::ok` to get the receipt claim, assuming the execution ended normally (e.g. did not panic) which covers every use case so far.